src/engine/renderer/gl_shader.cpp

```diff
@@ -2113,71 +2085,74 @@ uint32_t* GLUniform::WriteToBuffer( uint32_t* buffer ) {
+	for ( uint32_t i = 0; i < uniforms.size(); i++ ) {
+		if ( uniforms[i]->GetSTD430Size() <= alignment ) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
 // Compute std430 size/alignment and sort uniforms from highest to lowest alignment
 void GLShader::PostProcessUniforms() {
 	if ( !_useMaterialSystem ) {
 		return;
 	}
 
-	std::vector<GLUniform*> tmp;
-
-	std::vector<GLUniform*> globalUniforms;
 	for ( GLUniform* uniform : _uniforms ) {
-		if ( uniform->IsGlobal() || uniform->IsTexture() ) {
-			globalUniforms.emplace_back( uniform );
+		if ( !uniform->IsGlobal() ) {
+			_materialSystemUniforms.emplace_back( uniform );
 		}
 	}
-	for ( GLUniform* uniform : globalUniforms ) {
-		_uniforms.erase( std::remove( _uniforms.begin(), _uniforms.end(), uniform ), _uniforms.end() );
-	}
+
+	std::sort( _materialSystemUniforms.begin(), _materialSystemUniforms.end(),
+		[]( const GLUniform* lhs, const GLUniform* rhs ) {
+			return lhs->GetSTD430Size() > rhs->GetSTD430Size();
+		}
+	);
 
 	// Sort uniforms from highest to lowest alignment so we don't need to pad uniforms (other than vec3s)
-	const uint numUniforms = _uniforms.size();
-	GLuint structSize = 0;
+	const uint numUniforms = _materialSystemUniforms.size();
+	std::vector<GLUniform*> tmp;
 	while ( tmp.size() < numUniforms ) {
 		// Higher-alignment uniforms first to avoid wasting memory
-		GLuint highestAlignment = 0;
-		int highestUniform = 0;
-		for( uint i = 0; i < _uniforms.size(); i++ ) {
-			if ( _uniforms[i]->GetSTD430Alignment() > highestAlignment ) {
-				highestAlignment = _uniforms[i]->GetSTD430Alignment();
-				highestUniform = i;
-			}
-
-			if ( highestAlignment == 4 ) {
-				break; // 4-component is the highest alignment in std430
-			}
+		GLuint size = _materialSystemUniforms[0]->GetSTD430Size();
+		GLuint components = _materialSystemUniforms[0]->GetComponentSize();
+		size = components ? PAD( size, 4 ) * components : size;
+		GLuint alignmentConsume = 4 - size % 4;
+
+		GLUniform* tmpUniform = _materialSystemUniforms[0];
+		tmp.emplace_back( _materialSystemUniforms[0] );
+		_materialSystemUniforms.erase( _materialSystemUniforms.begin() );
+
+		int uniform;
+		while ( ( alignmentConsume & 3 ) && _materialSystemUniforms.size()
+			&& ( uniform = FindUniformForAlignment( _materialSystemUniforms, alignmentConsume ) ) != -1 ) {
+			alignmentConsume -= _materialSystemUniforms[uniform]->GetSTD430Size();
+
+			tmp.emplace_back( _materialSystemUniforms[uniform] );
+			_materialSystemUniforms.erase( _materialSystemUniforms.begin() + uniform );
 		}
-			
-		const GLuint size = _uniforms[highestUniform]->GetSTD430Size();
-		if ( _uniforms[highestUniform]->GetComponentSize() != 0 ) {
-			structSize += ( size == 3 ) ? 4 * _uniforms[highestUniform]->GetComponentSize() :
-												size * _uniforms[highestUniform]->GetComponentSize();
-		} else {
-			structSize += ( size == 3 ) ? 4 : size;
+
+		if ( alignmentConsume ) {
```

Comment:
This `while` followed by `if` structure doesn't really make sense since `SetSTD430SizeToAlignment()` would only work correctly if the while loop ran 0 times. `if` (uniform with exactly the needed alignment found) insert it `else` set padding, would make more sense